File: Solutions/104-Maximum Depth of Binary Tree/104-Maximum Depth of Binary Tree.js
//Fourth fastest
var maxDepth = function(root) {
    if(!root) return 0;
    let maxDepth = 1; // 
    
    const dfs = (root, depth) => {
        if(root) depth++;
        if(root.left) {
          dfs(root.left, depth);  
        } 
        if(root.right) {
            dfs(root.right, depth);
        }
        maxDepth = Math.max(maxDepth, depth)
    }
    
    dfs(root, 0);
    return maxDepth;
};

//Third fastest
var maxDepth = function(root) {
    // input validation
    if (!root) return 0;
    
    return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
};

//Second fastest
var maxDepth = function(root) {
    
    if(root === null) return 0;
    
    return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
};

//Fastest
var maxDepth = function(root) {
  if(!root) {
      return 0;
  }  
    return 1 + Math.max(maxDepth(root.left),maxDepth(root.right));
};

File: Solutions/104-Maximum Depth of Binary Tree/104-Maximum Depth of Binary Tree.rb
def max_depth(root)
    return 0 if root==nil
    [1+max_depth(root.left),1+max_depth(root.right)].max
end

#Third fastest
def max_depth(root)
    return 0 if root.nil?
  
    1 + [max_depth(root.left), max_depth(root.right)].max
  end

#Second fastest
def max_depth(root)
    unless root
        return 0
    else
        [1+max_depth(root.left),1+max_depth(root.right)].max
    end
end

#Fastest
Result = Struct.new(:data)

def max_depth(root)
    return 0 if root.nil?
    [max_depth(root.left), max_depth(root.right)].max + 1
end